packages/playwright-test/src/runner.ts

```diff
@@ -378,146 +383,132 @@ export class Runner {
-    if (!rootSuite.allTests().length && !options.passWithNoTests)
-      this._fatalErrors.push(createNoTestsError());
+    // Load tests.
+    let status: FullResult['status'] = 'passed';
+    let loadedTests!: { rootSuite: Suite, testGroups: TestGroup[] };
+    taskRunner.addTask('load tests', async () => {
+      // Each entry is an array of test groups that can be run concurrently. All
+      // test groups from the previos entries must finish before entry starts.
+      loadedTests = await this._loadAllTests(options);
+      if (this._reporter.hasErrors) {
+        status = 'failed';
+        taskRunner.stop();
+        return;
+      }
+    });
 
-    this._filterForCurrentShard(rootSuite, testGroups);
+    if (!options.listOnly) {
+      taskRunner.addTask('prepare to run', async () => {
+        // Fail when no tests.
+        if (!loadedTests.rootSuite.allTests().length && !options.passWithNoTests) {
+          this._reporter.onError(createNoTestsError());
+          status = 'failed';
+          taskRunner.stop();
+          return;
+        }
 
-    config._maxConcurrentTestGroups = testGroups.length;
+        // Remove output directores.
+        if (!await this._removeOutputDirs(options)) {
+          status = 'failed';
+          taskRunner.stop();
+          return;
+        }
+      });
 
-    const result: FullResult = { status: 'passed' };
-    for (const plugin of this._plugins)
-      await plugin.begin?.(rootSuite);
+      taskRunner.addTask('plugin begin', async () => {
+        for (const plugin of this._plugins)
+          await plugin.begin?.(loadedTests.rootSuite);
+      });
+    }
 
-    // Report begin
-    this._reporter.onBegin?.(config, rootSuite);
+    taskRunner.addTask('report begin', async () => {
+      this._reporter.onBegin?.(config, loadedTests.rootSuite);
 
-    // Bail out on errors prior to running global setup.
-    if (this._fatalErrors.length) {
-      for (const error of this._fatalErrors)
-        this._reporter.onError?.(error);
-      return { status: 'failed' };
-    }
+      return async () => {
+        await this._reporter.onEnd?.({ status });
+      };
+    });
 
-    // Bail out if list mode only, don't do any work.
-    if (options.listOnly)
-      return { status: 'passed' };
-
-    // Remove output directores.
-    if (!await this._removeOutputDirs(options))
-      return { status: 'failed' };
-
-    if (config._ignoreSnapshots) {
-      this._reporter.onStdOut?.(colors.dim([
-        'NOTE: running with "ignoreSnapshots" option. All of the following asserts are silently ignored:',
-        '- expect().toMatchSnapshot()',
-        '- expect().toHaveScreenshot()',
-        '',
-      ].join('\n')));
-    }
+    if (!options.listOnly) {
+      taskRunner.addTask('test suite', async () => {
+        const { rootSuite, testGroups } = loadedTests;
+        this._filterForCurrentShard(rootSuite, testGroups);
+        config._maxConcurrentTestGroups = testGroups.length;
```

Comment:
I think this should be done before `reporter.onBegin()`, since it's a part of `Running x tests on y workers`.

Reply:
done